File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = list()
        num = list()
        n = list()

        for letter in s:
            if letter.isdigit(): n.append(letter)
            elif letter == ']':
                target = list()
                while stack and stack[-1] != '[':
                    target.append(stack.pop())
                stack.pop()
                for i in range(num.pop()):
                    for j in range(len(target) - 1, -1, -1):
                        stack.append(target[j])
            elif letter == '[':
                stack.append(letter)
                num.append(int(''.join(n)))
                n.clear()
            else:
                stack.append(letter)

        return ''.join(stack)

File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {
        StringBuilder sb = new StringBuilder();
        Stack<Integer> num = new Stack();
        int n = 0;

        for (int i = 0; i < s.length(); i++) {
            if (Character.isDigit(s.charAt(i))) {
                n = n * 10 + s.charAt(i) - '0';
            } else if (s.charAt(i) == ']') {
                StringBuilder temp = new StringBuilder();
                int limit = num.pop();
                while (!sb.isEmpty() && sb.charAt(sb.length() - 1) != '[') {
                    temp.append(sb.charAt(sb.length() - 1));
                    sb.deleteCharAt(sb.length() - 1);
                }
                sb.deleteCharAt(sb.length() - 1);
                temp.reverse();
                for (int j = 0; j < limit; j++) {
                    sb.append(temp.toString());
                }
            } else if (s.charAt(i) == '[') {
                sb.append(s.charAt(i));
                num.add(n);
                n = 0;
            } else {
                sb.append(s.charAt(i));
            }
        }

        return sb.toString();
    }
}